Shiming-PCM applied on SETH model

ShimingYIN opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
This work is firstly to apply static_PCM on SETH model, which uses the cladding temperature of several timesteps in SETH-C to predict temperature of one timestep in SETH-D. Secondly, it applies snapshot_PCM on SETH model, which uses the temperature of each timestep in SETH-C to predict temperature of relevant timestep in SETH-D. The relevant RAVEN input files and test dataset is included here. 

##### What are the significant changes in functionality due to this change request?
It adds snapshot_PCM function in RAVEN.

moosebuild commented:
Job [Test Ubuntu 18 PIP](https://civet.inl.gov/job/1240211/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Ubuntu 18 PIP](https://civet.inl.gov/job/1240211/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test CentOS 7](https://civet.inl.gov/job/1240205/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test CentOS 8](https://civet.inl.gov/job/1240206/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Fedora 31](https://civet.inl.gov/job/1240207/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Fedora 32](https://civet.inl.gov/job/1240208/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Ubuntu 16](https://civet.inl.gov/job/1240209/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Ubuntu 18-2 Python 3](https://civet.inl.gov/job/1240210/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Ubuntu 20-2 Optional](https://civet.inl.gov/job/1240212/) on 8fcae41 : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Test Ubuntu 18-2 Python 3](https://civet.inl.gov/job/1240210/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test Ubuntu 20-2 Optional](https://civet.inl.gov/job/1240212/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test CentOS 8](https://civet.inl.gov/job/1240206/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test Fedora 31](https://civet.inl.gov/job/1240207/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test Fedora 32](https://civet.inl.gov/job/1240208/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test Ubuntu 16](https://civet.inl.gov/job/1240209/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test mac](https://civet.inl.gov/job/1240202/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

moosebuild commented:
Job [Test Ubuntu 18 PIP](https://civet.inl.gov/job/1240211/) on 8fcae41 : invalidated by @joshua-cogliati-inl

mystery failure

wangcj05 commented:
@ShimingYIN Could you try to reduce the size of the test? Can you reduce the number of realizations and time steps? or Can you design a simple test case? It is hard for us to maintain such large test case.

ShimingYIN commented:
> @ShimingYIN Could you try to reduce the size of the test? Can you reduce the number of realizations and time steps? or Can you design a simple test case? It is hard for us to maintain such large test case.

@wangcj05 Yes, I will reduce the number of realizations to make it simple and will update it later. On my laptop, this example took 13 mins to be finished. Sorry, I didn't know it was too big for a test.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping_snapshot.py:
```diff
@@ -0,0 +1,243 @@
+from .. import ValidationBase
+#Internal Modules End--------------------------------------------------------------------------------
+
+class PhysicsGuidedCoverageMapping_snapshot(ValidationBase):
```
@ShimingYIN Please rework on this PR. The class "PhysicsGuidedCoverageMapping_snapshot" should inherit from "PhysicsGuidedCoverageMapping" or it should be combined with existing "PhysicsGuidedCoverageMapping" class. If  the updates mainly focus on the loop over time, I would suggest to integrate this class with "PhysicsGuidedCoverageMapping" instead to write a new class. Please let me know if you have any question.

ShimingYIN replied:
Understood. I'll modify this problem on class inheritance and pull the request again

joshua-cogliati-inl commented on ravenframework/contrib/PythonFMU/pythonfmu.egg-info/PKG-INFO:
```diff
@@ -0,0 +1,184 @@
+Metadata-Version: 2.1
```
This file should not be commited.

joshua-cogliati-inl replied:
Fixed.

wangcj05 commented:
@ShimingYIN Could you resolve all the comments? I will try to review it before our meeting.

joshua-cogliati-inl commented on ravenframework/contrib/PythonFMU/pythonfmu.egg-info/SOURCES.txt:
```diff
@@ -0,0 +1,86 @@
+LICENSE
```
Specifically, the entire pythonfmu.egg-info directory should not be committed.  (I forgot to put this in the .gitignore, but the .gitignore has since been updated)

ShimingYIN replied:
@joshua-cogliati-inl Thanks for the comments. I have modified the committed content.

joshua-cogliati-inl replied:
Thanks for fixing :)

joshua-cogliati-inl commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/.ravenStatus:
```diff
@@ -0,0 +1 @@
+Success
```
.DS_Store and .ravenStatus should not be added to the repository.

ShimingYIN replied:
> @ShimingYIN Could you resolve all the comments? I will try to review it before our meeting.

@wangcj05 Sorry for the delay. I have removed the class "PhysicsGuidedCoverageMapping_snapshot" and combined it with the class "PhysicsGuidedCoverageMapping". And the SETH data applied for snapshot_PCM has been reduced, thus the test time is around 30sec now. The '.DS_Store' and '.ravenStatus' files have been removed and 'pythonfmu.egg-info' directory is not commited now.

joshua-cogliati-inl replied:
Thanks for fixing.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -88,10 +90,23 @@ def run(self, inputIn):
-    evaluation ={k: np.atleast_1d(val) for k, val in self._evaluate(dataDict, **{'dataobjectNames': names}).items()}
+
+    if 'time' in inputIn['Data'][0][-1].coords.keys():
+      PCMVersion = 'Snapshot'
```
FYI, the code standard for RAVEN is "camelBack", could you change "PCMVersion" to "pcmVersion"?

ShimingYIN replied:
The variable name has been corrected.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -88,10 +90,23 @@ def run(self, inputIn):
+
+    if 'time' in inputIn['Data'][0][-1].coords.keys():
+      PCMVersion = 'Snapshot'
+      print('Running Snapshot-PCM')
```
Please change all print to self.raiseAMessage

ShimingYIN replied:
The printing function has been modified and moved to _evaluate function.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -88,10 +90,23 @@ def run(self, inputIn):
+    if 'time' in inputIn['Data'][0][-1].coords.keys():
+      PCMVersion = 'Snapshot'
+      print('Running Snapshot-PCM')
+      pivotParameter = self.pivotParameter
+      if len(inputIn['Data'][0][-1].indexes) and self.pivotParameter is None:
+        if 'dynamic' not in self.dynamicType: #self.model.dataType:
+          self.raiseAnError(IOError, "The validation algorithm '{}' is not a dynamic model but time-dependent data has been inputted in object {}".format(self._type, inputIn['Data'][0][-1].name))
+      evaluation ={k: np.atleast_1d(val) for k, val in self._evaluate(PCMVersion, dataDict, **{'dataobjectNames': names}).items()}
+      evaluation[pivotParameter] = inputIn['Data'][0][-1]['time']
+    else:
+      PCMVersion = 'Static'
+      print('Running Static-PCM')
+      evaluation ={k: np.atleast_1d(val) for k, val in self._evaluate(PCMVersion, dataDict, **{'dataobjectNames': names}).items()}
```
I think you can use self.dynamic variable to indicate whether it is snapshot or static, and I would suggest you do not modify the function API for _evaluate method, since you can use self.dynamic in the _evaluate method to indicate the pcm version

ShimingYIN replied:
The API function for _evaluate method has been simplified. As snapshot PCM needs pivotParameter, an IF statement to obtain the values for pivotParameter is kept.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -103,125 +118,251 @@ def _evaluate(self, datasets, **kwargs):
+        # Condition number of matrix of feature and target
+        condNum = np.linalg.cond(vals)
+        # If condition number is greater than 100
+        invErr = 100
+        # Check whether the covavariance matrix is positive definite
+        if condNum>=invErr:
+          # If singular matrix, measurement of Experiment is directly transfered
+          # as predicted Application
+          pdfAppPred = knlMsr(Y[0, :])
+        else:
+          # If not, KDE of Experiment and Application
+          knl = stats.gaussian_kde(vals)
+          # Joint PDF of Experiment and Application
+          Z = np.reshape(knl(psts).T, X.shape)
+          # yAppPred by integrating p(yexp, yapp)p(ymsr) over [yexp.min(), yexp.max()]
+          pdfAppPred = np.dot(Z, pdfMsr.reshape(pdfMsr.shape[0], 1))
+
+        # Normalized PDF of predicted application
+        pdfAppPredNorm = pdfAppPred.flatten()/pdfAppPred.sum()/np.diff(Y[0, :])[0]
+
+        # Calculate Expectation (average value) of predicted application
+        # by integrating xf(x), where f(x) is PDF of x
+        predMean = 0.0
+        for i in range(len(Y[0, :])):
+          predMean += Y[0, i]*pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+        # Calculate Variance of predicted application
+        # by integrating (x-mu_x)^2f(x), where f(x) is PDF of x
+        predVar = 0.0
+        for i in range(len(Y[0, :])):
+          predVar += (Y[0, i]-predMean)**2.0 * pdfAppPredNorm[i]*(Y[0, 1]-Y[0, 0])
+
+        # Predicted standard deviation is square root of variance
+        predStd = np.sqrt(predVar)
+        # Prior standard deviation is the sample standard deviation
+        # Consider probability weights in the future
+        priStd = np.std(yAppStd)
+        # Uncertainty reduction fraction is 1.0-sigma_pred/sigma_pri
+        name = "pri_post_stdReduct_" + targ.split('|')[-1]
+        outputDict[name] = (1.0-predStd/priStd)
```
please make a function for these lines, since you can reuse it in the snapshot method

ShimingYIN replied:
The _evaluate method has been simplified.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -103,125 +118,251 @@ def _evaluate(self, datasets, **kwargs):
 
-      # Single Experiment response
-      if yExpStd.shape[1]==1:
+      for feat, msr, targ in zip(self.features, self.measurements, self.targets):
```
why self.targets only appears in snapshot method, not both?

ShimingYIN replied:
> I have several comments that I have added in the PR. In addition, please check the following:
> 
> 1. check all the csv files, if they are not used, please remove the file
> 2. It seems there is no results for snapshot test
> 3. You need to update tests file in your test folder in order your added tests can be tested. (I can show you how to do it.) Here is the example:
>    ![image](https://user-images.githubusercontent.com/7321071/211610591-a8b1b47d-8103-4cb2-8c30-3cd2c9cde36c.png)
> 
> The csv files need to be placed in the gold folder ![image](https://user-images.githubusercontent.com/7321071/211610767-83f6384e-6e08-4195-aff8-0cf1e8731540.png)

@wangcj05  Thanks for the comments Congjian. According to the suggestions: 
1. The unused data in StaticExample folder has been removed. The 877 csv files in SnapshotExample folder are are the 877 samples and all needed for test.
2. The results for snapshot test are in file 'pcmSnapshot_metric_dump_0.csv', which has been moved to gold folder.
3. The 'tests' file has been updated with static/snapshot PCM test info.

ShimingYIN replied:
> why self.targets only appears in snapshot method, not both?

It has been included in static method previously. Now I have simplified the _evaluate function and combine the data reading for static/snapshot PCM.

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/pcm_metric_dump.xml:
```diff
@@ -0,0 +1,10 @@
+<DataObjectMetadata name="pcm_metric">
```
please remove this file

ShimingYIN replied:
File has been deleted.

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/pcm_metric_dump.csv:
```diff
@@ -0,0 +1,2 @@
+pri_post_stdReduct_time900
```
This file need to be moved to gold folder

ShimingYIN replied:
Output files have been moved to gold folder.

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SETH.xml:
```diff
@@ -0,0 +1,85 @@
+        The output name convention is 'pri_post_stdReduct_'+"Target name".
+    </description>
+    <revisions>
+      <revision author="syin" date="2022-10-03">...</revision>
```
please update this if you have revisions

ShimingYIN replied:
The revision date has been modified and this test file name has been changed according to CamelBack format.

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm_snapshotSETH.xml:
```diff
@@ -0,0 +1,97 @@
+        and has two columns:'time' and 'pri_post_stdReduct'.
+    </description>
+    <revisions>
+      <revision author="syin" date="2022-12-06">...</revision>
```
please update this, otherwise remove it

ShimingYIN replied:
The revision date has been modified and this test file name has been changed according to CamelBack format.

wangcj05 commented:
@ShimingYIN I have several comments for you PR, please try to address it before the meeting.

ShimingYIN commented:
> @ShimingYIN I have several comments for you PR, please try to address it before the meeting.

@wangcj05  Hi Congjian, I found the updated request has failed GitHub tests, but the SETH examples have been successfully run on my laptop and the 'delete_trailing_whitespace' command has been applied before pulling the request. Previously when I did these, the Github tests all passed, but it didn't work this time. I haven't found a good solution online, and wonder if you know the reason why the tests succeed locally but fail remotely.

wangcj05 commented:
Failures reasons: 
<img width="1113" alt="image" src="https://user-images.githubusercontent.com/7321071/212753223-ddf490e0-aa02-4820-bd4c-14df78dd5e46.png">



```
 ... successfully imported "ExamplePlugin" ...
WARNING: "/home/civet/civet/build_0/raven/ravenframework/contrib/pp" already in system path. Skipping CPP setup
Traceback (most recent call last):
  File "/home/civet/civet/build_0/raven/raven_framework.py", line 26, in <module>
    sys.exit(main(True))
  File "/home/civet/civet/build_0/raven/ravenframework/Driver.py", line 127, in main
    tree = TS.parse(open(inputFile,'r'))
FileNotFoundError: [Errno 2] No such file or directory: '/home/civet/civet/build_0/raven/tests/framework/PostProcessors/Validation/test_validation_gate_pcm_StaticSeth.xml'

Running test failed with exit code 1
(409F2/833) Failed (  2.54sec)tests/framework/PostProcessors/Validation/test_validation_gate_pcm_StaticSeth
Output of'tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth':
```

```
 ... successfully imported "ExamplePlugin" ...
WARNING: "/home/civet/civet/build_0/raven/ravenframework/contrib/pp" already in system path. Skipping CPP setup
Traceback (most recent call last):
  File "/home/civet/civet/build_0/raven/raven_framework.py", line 26, in <module>
    sys.exit(main(True))
  File "/home/civet/civet/build_0/raven/ravenframework/Driver.py", line 127, in main
    tree = TS.parse(open(inputFile,'r'))
FileNotFoundError: [Errno 2] No such file or directory: '/home/civet/civet/build_0/raven/tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth.xml'

Running test failed with exit code 1
(410F3/833) Failed (  2.55sec)tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth
```

wangcj05 commented:
@ShimingYIN Two of the files are not available

wangcj05 commented:
The other failure seems to be pcm_metric is not calculated and formatted correctly.

wangcj05 commented on tests/framework/PostProcessors/Validation/tests:
```diff
@@ -27,4 +27,18 @@
   [../]
+  [./test_validation_gate_pcm_StaticSeth]
+    type = 'RavenFramework'
+    input = 'test_validation_gate_pcm_StaticSeth.xml'
```
The file is not added in your branch, please try to add it.

ShimingYIN commented:
> ```python
> t failed with exit code 1
> ```

Thanks Congjian. I'll fix this and test again.

moosebuild commented:
Job [Test qsubs sawtooth](https://civet.inl.gov/job/1477344/) on 76fba7e : invalidated by @joshua-cogliati-inl

restarted civet

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1508511/) on 818b254 : invalidated by @joshua-cogliati-inl

failed in fetch

wangcj05 commented:
@ShimingYIN FYI, the following two tests are failed:

```
/home/civet/.conda/envs/raven_libraries/lib/python3.10/site-packages/sklearn/linear_model/_base.py:133: FutureWarning: The default of 'normalize' will be set to False in version 1.2 and deprecated in version 1.4.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), OrthogonalMatchingPursuit())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)


  warnings.warn(
/home/civet/.conda/envs/raven_libraries/lib/python3.10/site-packages/sklearn/linear_model/_base.py:133: FutureWarning: The default of 'normalize' will be set to False in version 1.2 and deprecated in version 1.4.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), OrthogonalMatchingPursuit())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)


  warnings.warn(
/home/civet/.conda/envs/raven_libraries/lib/python3.10/site-packages/sklearn/linear_model/_base.py:133: FutureWarning: The default of 'normalize' will be set to False in version 1.2 and deprecated in version 1.4.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), OrthogonalMatchingPursuit())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)


  warnings.warn(
(    1.63 sec) STEP SINGLERUN           : Message         -> ***       Run finished        ***
(    1.63 sec) STEP SINGLERUN           : Message         -> ***     Closing the step      ***
(    1.63 sec) STEP SINGLERUN           : Message         -> ***        Step closed        ***
(    1.63 sec) SIMULATION               : Message         -> -- End step PCM of type: PostProcess --

(    1.64 sec) SIMULATION               : Message         -> Run complete!




DIFF in tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/pcm_metric_dump.csv: 
  Different values in static_pri_post_stdReduct_F2 for 0.832328312741 and 0.236223469749
  Different values in static_pri_post_stdReduct_F3 for 0.832328312741 and 0.236223469749
  Different values in static_pri_post_stdReduct_F4 for 0.832328312741 and 0.236223469749
  | Difference | statistics:
    MEAN    diff.: 2.523478483e+00
    LARGEST diff.: 2.523478483e+00
(412F1/852) Diff   (  3.73sec)tests/framework/PostProcessors/Validation/test_validation_gate_pcm
Output of'tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth':

Copyright 2017 Battelle Energy Alliance, LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
  

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

      .---.        .------######       #####     ###   ###  ########  ###    ###
     /     \  __  /    --###  ###    ###  ###   ###   ###  ###       #####  ###
    / /     \(  )/    --###  ###    ###   ###  ###   ###  ######    ### ######
   //////   ' \/ `   --#######     #########  ###   ###  ###       ###  #####
  //// / // :    :   -###   ###   ###   ###    ######   ####      ###   ####
 // /   /  /`    '---###    ###  ###   ###      ###    ########  ###    ###
//          //..\\
===========UU====UU=============================================================
           '//||\\`
             ''``
    
RAVEN Python dependencies located and checked.
Loading plugin "TEAL" at /home/civet/civet/build_0/raven/plugins/TEAL
 ... successfully imported "TEAL" ...
Loading plugin "HERON" at /home/civet/civet/build_0/raven/plugins/HERON
 ... successfully imported "HERON" ...
Loading plugin "SR2ML" at /home/civet/civet/build_0/raven/plugins/SR2ML
 ... successfully imported "SR2ML" ...
Loading plugin "LOGOS" at /home/civet/civet/build_0/raven/plugins/LOGOS
 ... successfully imported "LOGOS" ...
Loading plugin "SRAW" at /home/civet/civet/build_0/raven/plugins/SRAW
 ... successfully imported "SRAW" ...
Loading plugin "FARM" at /home/civet/civet/build_0/raven/plugins/FARM
 ... successfully imported "FARM" ...
Loading plugin "ExamplePlugin" at /home/civet/civet/build_0/raven/plugins/ExamplePlugin
 ... successfully imported "ExamplePlugin" ...
WARNING: "/home/civet/civet/build_0/raven/ravenframework/contrib/pp" already in system path. Skipping CPP setup
Traceback (most recent call last):
  File "/home/civet/civet/build_0/raven/raven_framework.py", line 25, in <module>
    sys.exit(main(True))
  File "/home/civet/civet/build_0/raven/ravenframework/Driver.py", line 127, in main
    tree = TS.parse(open(inputFile,'r'))
FileNotFoundError: [Errno 2] No such file or directory: '/home/civet/civet/build_0/raven/tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth.xml'

Running test failed with exit code 1
```

wangcj05 commented:
@ShimingYIN Please try to reduce the number and size of the csv files that you are using. Regression test should be simple.

wangcj05 commented:
@ShimingYIN Have you resolved the issues for this PR?

ShimingYIN commented:
> @ShimingYIN Have you resolved the issues for this PR?

Sorry for checking the comments late, Congjian. I have simplified the input csv files for the tests and also try to fix the two failures you mentioned. Previously, tests on Snapshot_PCM and Static_PCM work well on my end, and I also committed the change of 'test_validation_gate_pcm_SnapshotSeth.xml', so I don't know how the failures come and will wait for the checks to be finished on the remote repository.

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/.DS_Store:
please remove this file

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/SnapshotExample/.DS_Store:
please remove this file

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/SnapshotExample/TSimpleGap_data.csv:
please reduce the size of this file

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/StaticExample/.DS_Store:
please remove this file

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/TimeDep/.DS_Store:
please remove this file

wangcj05 commented on tests/framework/PostProcessors/Validation/PhysicsGuidedCoverageMapping/data/TimeDep/TToptanGap_data.csv:
please reduce the size of this file

wangcj05 commented on tests/framework/PostProcessors/Validation/gold/PhysicsGuidedCoverageMapping/.DS_Store:
please remove this file

wangcj05 reviewed (changes requested):
@ShimingYIN Please check my comments, there are several files that need to be removed, and some file need to be reduced.

wangcj05 commented on tests/framework/PostProcessors/Validation/tests:
```diff
@@ -27,4 +27,25 @@
+  [../]
+  [./test_validation_gate_pcm_SnapshotSeth]
+    type = 'RavenFramework'
+    input = 'test_validation_gate_pcm_SnapshotSeth.xml'
```
The test name here is not right, please update it, please pay attention to the capital and lowercase letters

wangcj05 commented on tests/framework/PostProcessors/Validation/test_validation_gate_pcm_Snapshot.xml:
```diff
@@ -0,0 +1,94 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="all">
```
The filename is not as same as the one used in tests file, please check it @ShimingYIN 
Traceback (most recent call last):
  File "/home/civet/civet/build_0/raven/raven_framework.py", line 25, in <module>
    sys.exit(main(True))
  File "/home/civet/civet/build_0/raven/ravenframework/Driver.py", line 127, in main
    tree = TS.parse(open(inputFile,'r'))
FileNotFoundError: [Errno 2] No such file or directory: '/home/civet/civet/build_0/raven/tests/framework/PostProcessors/Validation/test_validation_gate_pcm_SnapshotSeth.xml'

wangcj05 commented:
@ShimingYIN Please try to resolve my comments, and please let me know if you have any question regarding my comments.

ShimingYIN commented:
> @ShimingYIN Please try to resolve my comments, and please let me know if you have any question regarding my comments.

Sorry, Congjian. The test failure results from my mistake of not committing the changes successfully. I have updated and added all the modified files there according to your previous comments.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1554173/) on 8fb7946 : invalidated by @wangcj05

ShimingYIN commented:
> @ShimingYIN Tests are now good, can you try to reduce the size of csv files?

@wangcj05 The sizes of .csv files have been further reduced and updates have been pushed to the remote.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -101,127 +111,280 @@ def _evaluate(self, datasets, **kwargs):
+      return outputArray
+
+    # functions for T-Dep PCM
+    def KNN(exp, app, expM, k=5, Metric='minkowski'):
```
please add docstring for this function

wangcj05 reviewed (changes requested):
@ShimingYIN I have a couple of comments for you to consider. Please let me know if you have any issue with my comments. In addition, please update the manual to reflect your changes.

ShimingYIN commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -88,10 +92,16 @@ def run(self, inputIn):
     dataDict = {self.getDataSetName(data): data for _, _, data in inputIn['Data']}
     names = list(dataDict.keys())#
-    evaluation ={k: np.atleast_1d(val) for k, val in self._evaluate(dataDict, **{'dataobjectNames': names}).items()}
+    evaluation ={k: np.atleast_1d(val) for k, val in self._evaluate(inputIn['Data'][0][-1].coords.keys(), dataDict, **{'dataobjectNames': names}).items()}
```
> Right now your are using "inputIn['Data'][0][-1].coords.keys()" to determine the type of PCM analysis, either 'static', 'snapshot' or 'Time dependent' analysis. In this case, the PCM post-processor only works for given data that include 'timeTdep' and 'timeSnapshot'. This will significantly limit the flexibility of this PostProcessor. Please define a new XML node in the RAVEN input file to indicate the type of analysis, and use that inside the code to perform corresponding analysis. Please also add checks for each analysis to make the right data is provided from the user.

Docstrings have been added. A new XML node to determine the PCM version has been applied and checks have been added to provide warnings if input data given by users are not in the required format.

ShimingYIN replied:
> Right now your are using "inputIn['Data'][0][-1].coords.keys()" to determine the type of PCM analysis, either 'static', 'snapshot' or 'Time dependent' analysis. In this case, the PCM post-processor only works for given data that include 'timeTdep' and 'timeSnapshot'. This will significantly limit the flexibility of this PostProcessor. Please define a new XML node in the RAVEN input file to indicate the type of analysis, and use that inside the code to perform corresponding analysis. Please also add checks for each analysis to make the right data is provided from the user.

Docstrings have been added. A new XML node to determine the PCM version has been applied and checks have been added to provide warnings if input data given by users are not in the required format.

ShimingYIN commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -101,127 +111,280 @@ def _evaluate(self, datasets, **kwargs):
+      for k in range(1,maxError.shape[0]+1):
+        res = x - u[:,:k]@np.diag(s)[:k,:k]@v_T[:k,:]#residual under different rank
+        maxError[k-1] = np.max(abs(res)) # maximum of the residual
+      xMAX=min(np.argwhere(maxError<error/10)) #rank corresponding to msr error
```
The prediction of pcmTdep is sensitive to the rank, which determines the dimension of U_exp, U_app subspaces for KNN regression. We will lose information under low rank, or overfit data with noise under high rank. 1/10 is decided here to ensure a rank corresponding to around 0.1% reconstruction error, because the error from SETH measurement data is around 2% which is too big for deciding on reconstruction error. I think I could make a new node in input XML file to make this error user-defined later, instead of calculated from measurement data.

wangcj05 replied:
Could you add a new xml node to control this?

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1595251/) on 2571ad8 : invalidated by @joshua-cogliati-inl

 Failed tests/framework/ROM/MSR/parallel

wangcj05 commented:
@ShimingYIN There is an issue with the test, it seems there is some file missing.
![image](https://github.com/idaholab/raven/assets/7321071/374be530-18b3-4e71-a1a6-e42e21bff827)

moosebuild commented:
All [jobs](https://civet.inl.gov/event/132936/) on d20d5d1 : canceled by @wangcj05

missing files

ShimingYIN commented:
> @ShimingYIN There is an issue with the test, it seems there is some file missing. ![image](https://user-images.githubusercontent.com/7321071/250623179-374be530-18b3-4e71-a1a6-e42e21bff827.png)

Sorry, Congjian, could you give the full screenshot as I couldn't see the full name of the missing file?

wangcj05 commented:
This file @ShimingYIN : 
![image](https://github.com/idaholab/raven/assets/7321071/1d81d3c8-911e-40cb-8998-d70233d0d896)

ShimingYIN commented:
> This file @ShimingYIN : ![image](https://user-images.githubusercontent.com/7321071/250626027-1d81d3c8-911e-40cb-8998-d70233d0d896.png)

Thanks Congjian, the missing file has been added.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/132978/) on 64c069d : canceled by @wangcj05

cancel for now

moosebuild commented:
All [jobs](https://civet.inl.gov/event/132978/) on 64c069d : invalidated by @wangcj05

Testing with raven library updates

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1631850/) on 64c069d : invalidated by @wangcj05

fetch issue

wangcj05 commented on doc/user_manual/PostProcessors/Validation.tex:
```diff
@@ -231,27 +231,37 @@ \subsubsection{Validation PostProcessors}
-
+\textbf{PCM} evaluates the uncertainty reduction fraction and obtain posterior distribution of Target 
+when using Feature(s) to validate each Target via Physics-guided Coverage Mapping (PCM) method. There are 
+three versions of PCM so far: 'Static', 'Snapshot', and 'Tdep'. Static PCM is for static problem, and Snapshot PCM 
```
it is better to use `' instead of ''

wangcj05 commented on doc/user_manual/PostProcessors/Validation.tex:
```diff
@@ -231,27 +231,37 @@ \subsubsection{Validation PostProcessors}
      measurements of the features. \nb Each measurement correspond to a feature listed in xml node \xmlNode{Features}. The
     number of measurements should be equal to the number of features and in the same order as the features listed in \xmlNode{Features}.
+  \item \xmlNode{pcmType}, \xmlDesc{string, required field}, contains the string given by users to choose the version 
+    of PCM to be applied. \nb It has three options: 'Static', 'Snapshot', and 'Tdep', corresponding to the three PCM versions.
```
better to use `' instead of ''

wangcj05 reviewed (changes requested):
@ShimingYIN Some minor comments. Please let me know if you can address them.

wangcj05 commented:
@ShimingYIN I have few more comments. Please let me know once you addressed them.

ShimingYIN commented:
> @ShimingYIN I have few more comments. Please let me know once you addressed them.

@wangcj05 Sorry for the delay, Congjian. I have just addressed the comments you made. An optional node 'ReconstructionError' has been added to determine the error corresponding to the rank of time series data, and the problem in user manual is solved.

wangcj05 commented on doc/user_manual/PostProcessors/Validation.tex:
```diff
@@ -249,7 +249,9 @@ \subsubsection{Validation PostProcessors}
   \item \xmlNode{pcmType}, \xmlDesc{string, required field}, contains the string given by users to choose the version 
-    of PCM to be applied. \nb It has three options: 'Static', 'Snapshot', and 'Tdep', corresponding to the three PCM versions.
+    of PCM to be applied. \nb It has three options: `Static', `Snapshot', and `Tdep', corresponding to the three PCM versions.
+  \item \xmlNode{ReconstructionError}, \xmlDesc{string, optional field}, contains the string given by users to determind the 
```
change 'string' to 'float'

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -55,6 +55,8 @@ class cls.
     specs.addSub(measurementsInput)
     pcmTypeInput = InputData.parameterInputFactory("pcmType", contentType=InputTypes.StringType)
     specs.addSub(pcmTypeInput)
+    recErrorInput = InputData.parameterInputFactory("ReconstructionError", contentType=InputTypes.StringListType)
```
change 'StringListType' to 'FloatType'

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -83,6 +85,11 @@ def _handleInput(self, paramInput):
+      elif child.getName() == 'ReconstructionError':
+        self.ReconstructionError = child.value
+      else :
+        self.ReconstructionError = None
```
assign default value instead

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -291,27 +298,33 @@ def FindRank(ref, x, xM):
       @ In, ref: array-like (1D), reference sample used for scaling data 'x'.
       @ In, x: array-like (2D), the matrix [samples x timesteps] for which the rank is to be determined.
-      @ In, xM: array-like (2D), the matrix [samples x timesteps] used to calculate measurement error.
+      @ In, recError, array-like (1D), reconstruction error to determine the rank of time series data.
```
change array-like (1D) to float

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -291,27 +298,33 @@ def FindRank(ref, x, xM):
+      # U_exp, U_app subspaces for KNN regression. We will lose information under low rank, or
+      # overfit data with noise under high rank.
+      # This error can be made as user-defined.
+      if recError == None:
+        recError = 0.001 # 0.001 is decided as we want a rank corresponding to around 0.1% reconstruction error.
+      else:
+        recError = float(recError[0])
```
Once you made the changes according to my previous comments, you do not need this lines, since recError becomes float instead of StringList.

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -321,6 +334,7 @@ def pcmTdep(featData, msrData, targData):
         by the estimated application coefficients.
 
         @ In, featData, msrData,, targData : array-like (2D), the input data for features, measurements, and targets.
+        @ In, recError, array-like (1D), reconstruction error to determine the rank of time series data.
```
change array-like to float

wangcj05 commented on ravenframework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py:
```diff
@@ -399,6 +413,12 @@ def pcmTdep(featData, msrData, targData):
 
     pcmVersion = self.pcmType
 
+
+    if self.ReconstructionError != None:
+      recError = self.ReconstructionError #reconstruction error to determine the rank of time series data.
+    else:
+      recError = None
+
```
There is a default for self.ReconstructionError, so you do not need the if else check list.

wangcj05 reviewed (changes requested):
@ShimingYIN Thanks for your changes, please revise your recent commits according to my comments.

ShimingYIN commented:
> @ShimingYIN Thanks for your changes, please revise your recent commits according to my comments.

@wangcj05 Thanks for the comments, Congjian. The parts mentioned have been modified.

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1653259/) on 40ef2b9 : invalidated by @wangcj05

segmentation fault

moosebuild commented:
Job [Mingw Test](https://civet.inl.gov/job/1653259/) on 40ef2b9 : invalidated by @wangcj05

fetch issue probably due to HPC issue

wangcj05 commented:
Changes are good, checklist is satisfied. PR can be merged.

wangcj05 commented:
@ShimingYIN Thanks for your contribution. I have merged your PR.
